src/transformers/optimization.py

```diff
@@ -940,6 +943,325 @@ def get_lr(self):
+
+                    state["vmean"].mul_(beta2).add_(tmp_lr, alpha=1 - beta2)
+                    v = state["vmean"]
+
+                    state["iteration"] += 1
+                    if group["weight_decay"] != 0:
+                        p.data.mul_(1 - lr * group["weight_decay"])
+
+                    state["m"].lerp_(grad, 1 - beta1)
+
+                    bias_correction_1 = 1 - beta1 ** state["iteration"]
+                    bias_correction_2 = 1 - beta2 ** state["iteration"]
+                    bias_correction_2_sqrt = math.sqrt(bias_correction_2)
+
+                    h = (v.sqrt() / bias_correction_2_sqrt).add_(eps)
+                    stepsize = ((1 / bias_correction_1) / h).view(head, 1)
+
+                    update = state["m"] * (stepsize.to(state["m"].device))
+
+                    if p.dim() > 1:
+                        d0, d1 = p.size()
+                        update = update.view(d0, d1)
+                    else:
+                        update = update.view(-1)
+
+                    update.mul_(lr)
+                    p.add_(-update)
+
+                else:
+                    if len(state) == 0:
+                        dimension = torch.tensor(p.data.numel()).to(self.device).to(torch.float32)
+                        reduced = False
+                        if (self.world_size > 1) and (self.model_sharding is True):
+                            tensor_list = [torch.zeros_like(dimension) for _ in range(self.world_size)]
+
+                            dist.all_gather(tensor_list, dimension)
+                            s = 0
+                            dimension = 0
+                            for d in tensor_list:
+                                if d > 0:
+                                    s = s + 1
+                                dimension = dimension + d
+                            if s >= 2:
+                                reduced = True
+
+                        state["m"] = torch.zeros_like(p.data).to(torch.float32)
+                        state["iteration"] = 0
+                        state["reduced"] = reduced
+
+                        state["vmean"] = torch.zeros_like(torch.sum(p.data * p.data)).to(self.device)
+                        state["dimension"] = dimension.item()
+                    if p.grad is None:
+                        tmp_lr = torch.zeros_like(torch.sum(p.data * p.data)).to(self.device)
+                    else:
+                        grad = p.grad.data.to(torch.float32)
+                        tmp_lr = torch.sum(grad * grad).to(self.device)
+
+                    if state["reduced"]:
+                        if "device_mesh" in dir(tmp_lr):
+                            lr_local = tmp_lr.to_local()
+                            dist.all_reduce(lr_local, op=dist.ReduceOp.SUM)
+                            tmp_lr.redistribute(placements=[Replicate()])
+                        else:
+                            dist.all_reduce(tmp_lr, op=dist.ReduceOp.SUM)
```

Comment:
Similarly

Reply:
The aforementioned changes have been completed. Kindly review them.